traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go

```diff
@@ -0,0 +1,389 @@
+	if err != nil {
+		log.Errorf("received error: %++v from delete execution", err)
+		return tc.DBError, tc.SystemError
+	}
+
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		log.Errorf("getting rows affected: %v", err)
+		return tc.DBError, tc.SystemError
+	}
+
+	if rowsAffected != 1 {
+		if rowsAffected < 1 {
+			return errors.New("no federation with that id found"), tc.DataMissingError
+		} else {
+			return fmt.Errorf("this delete affected too many rows: %d", rowsAffected), tc.SystemError
+		}
+	}
+	return nil, tc.NoError
+}
+
+//Function not exported because although DELETE and UPDATE have normal tenacy check,
+//CREATE does not. No ds is associated on create. This isn't used for READ because
+//psql doesn't like nested queries.
+func (fed TOCDNFederation) isTenantAuthorized() (bool, error) {
+	if fed.ID == nil {
+		log.Errorf("unexpected nil id\n")
+	}
+	//Note: the tenantID is not found via a recursive search. The recursive search
+	//is done in IsResourceAuthorizedToUser below.
+	tenantID, err := getTenantIDFromFedID(*fed.ID, fed.ReqInfo.Tx.Tx)
+	if err != nil {
+		//If nobody has claimed a tenant, that federation is publicly visible.
+		//This logically follows /federations/:id/deliveryservices
+		if err == sql.ErrNoRows {
+			log.Errorf("no tenacy")
+			return true, nil
+		}
+		log.Errorf("ran into error %v", err)
+		return false, err
+	}
+
+	//After IsResourceAuthorizedToUserTx is updated to no longer have `use_tenancy`,
+	//that will probably be better to use. For now, use the list.
+	list, err := tenant.GetUserTenantIDListTx(fed.ReqInfo.Tx.Tx, fed.ReqInfo.User.ID)
+	if err != nil {
+		return false, err
+	}
+	for _, id := range list {
+		if id == tenantID {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func getTenantIDFromFedID(id int, tx *sql.Tx) (int, error) {
+	tenantID := 0
+	query := `
+	SELECT tenant_id from federation
+	JOIN federation_deliveryservice as fd ON federation.id = fd.federation
+	JOIN deliveryservice as ds ON ds.id = fd.deliveryservice
+	WHERE federation.id = $1`
+	err := tx.QueryRow(query, id).Scan(&tenantID)
+	return tenantID, err
+}
+
+func selectByID() string {
+	return `
+	SELECT federation.id as id, cname, ttl, description, ds.id as ds_id, xml_id FROM federation
+	LEFT JOIN federation_deliveryservice as fd ON federation.id = fd.federation
+	LEFT JOIN deliveryservice as ds ON ds.id = fd.deliveryservice`
+	//WHERE federation.id = :id (determined by dbhelper)
+}
+
+func selectByCDNName() string {
+	return `
+	SELECT federation.id as id, cname, ttl, description, ds.id as ds_id, xml_id FROM federation
+	JOIN federation_deliveryservice as fd ON federation.id = fd.federation
+	JOIN deliveryservice as ds ON ds.id = fd.deliveryservice
+	JOIN cdn ON cdn.id = cdn_id`
+	//WHERE cdn.name = :cdn_name (determined by dbhelper)
+}
+
+func updateQuery() string {
+	return `
+	UPDATE federation SET
+	cname=:cname,
+	ttl=:ttl,
+	description=:description
+	WHERE id=:id
```

Comment:
I was following the [API] (http://traffic-control-cdn.readthedocs.io/en/latest/api/v12/federation.html), where last-updated isn't a part of the response. Since there were other mentions of last-updated from the same resource it seemed like an explicit decision to have it removed from this particular response. Are you saying I should add it now regardless?

Reply:
So a real problem we have as a project is that devs don't always update the documentation when they update the code. So "Code is King" and should be depended on more than the docs when rewriting existing Perl API endpoints.

Also, something that all our Go APIs are missing from the Perl version is returning all the same stuff on POST/PUT - typically JOINed fields like 'fooName' rather than just 'fooID'. We don't have that exact problem here, but I think we should still be returning `last_updated` here.